### ISSUE TITLE
should we use greedily decoded sentence for baseline reward computation?

### ISSUE THREAD

**Issue by Hannabrahman** [ruotianluo/self-critical.pytorch#142]
In the paper it is mentioned that the idea is to baseline the REINFORCE algorithm with  the  reward  obtained  by  the  current  model  under  the inference algorithm used at test time.
What if we want to use topk sampling at the test time? should we still use greedy decode output for baseline reward?

Thanks

**Comment by ruotianluo** [comment]
Model trained with self crtitical usually can't work with topk because the posterior is too peaky.

**Comment by Hannabrahman** [comment]
> Model trained with self crtitical usually can't work with topk because the posterior is too peaky.

@ruotianluo Thanks for you quick reply.
In this regard:
1) does it make sentence to have baseline reward using greedy, however during inference we use topk?
2) Do people usually use beam search during inference for a model trained by self critical?
In general what is the best approach if I want to use RL to finetune my model? Should I only use beam or greedy during inference?

**Comment by ruotianluo** [comment]
1. can't understand.
2. Yes.
3. Best approach of what?

**Comment by Hannabrahman** [comment]
>     1. can't understand.
> 
>     2. Yes.
> 
>     3. Best approach of what?

1- You mean if we train self-critical, using topk sampling during inference time won't work, right?
3- In other words, only beam search/greedy decoding (during inference) work with self-critical?

**Comment by ruotianluo** [comment]
Sorry for the ambiuty
1. Won't work means it will not generate very different captions because the posterior is too peaky.
3. Any decoding would work in terms of generating a caption; however, they may not work as you would expect.

You can certainly try,
